row_number variable description update

phidrho opened the pull request:
some users expect that `@row_number` variable respects what they currently see in attribute table so I'm proposing a better explanation for variable to make it clear

fix: #53536
fix: #54509

phidrho committed:
row_number variable description update

some users expect that `@row_number` variable respects what they currently see in attribute table so I'm proposing a better explanation for variable

mblesius commented:
Refs https://github.com/qgis/QGIS/issues/53536#issuecomment-2827002614

There are other places in QGIS where the `@row_number` variable is also available. It can be used when calculating param values and adding rows by expression. Here, `@row_number` actually returns the (internal) row number of the entries in the input table starting from 0. Thus, I would suggest to change the variable description, to make clear that (only?) for attribute tables, it will respect the order of features from the data source and not the current sort order in the attribute table.

![row_number_batch_processing](https://github.com/user-attachments/assets/8a3ef5e7-c11e-4a2d-b4f0-67129cc185b1)

phidrho committed:
added a note for indexing

in some cases indexing starts at 0, and in some cases indexing for same function starts at 1

phidrho commented:
> Refs [#53536 (comment)](https://github.com/qgis/QGIS/issues/53536#issuecomment-2827002614)
> 
> There are other places in QGIS where the `@row_number` variable is also available. It can be used when calculating param values and adding rows by expression. Here, `@row_number` actually returns the (internal) row number of the entries in the input table starting from 0. Thus, I would suggest to change the variable description, to make clear that (only?) for attribute tables, it will respect the order of features from the data source and not the current sort order in the attribute table.


@mblesius
Thank you for your comment. Please check if new version is good.

phidrho committed:
additional refinement

phidrho commented:
unstale

nyalldawson commented on src/core/expression/qgsexpression.cpp:
```diff
@@ -851,7 +851,7 @@ void QgsExpression::initVariableHelp()
   sVariableHelpTexts()->insert( QStringLiteral( "vector_tile_zoom" ), QCoreApplication::translate( "variable_help", "Exact vector tile zoom level of the map that is being rendered (derived from the current map scale). Normally in interval [0, 20]. Unlike @zoom_level, this variable is a floating point value which can be used to interpolate values between two integer zoom levels." ) );
 
-  sVariableHelpTexts()->insert( QStringLiteral( "row_number" ), QCoreApplication::translate( "variable_help", "Stores the number of the current row." ) );
+  sVariableHelpTexts()->insert( QStringLiteral( "row_number" ), QCoreApplication::translate( "variable_help", "Stores the number of the current row. For attribute table data it's respecting order of features from data source. If used from attribute table field calculator index starts at 1, otherwise (e.g. from processing tools) index starts from 0." ) );
```
```suggestion
  sVariableHelpTexts()->insert( QStringLiteral( "row_number" ), QCoreApplication::translate( "variable_help", "Stores the number of the current row." ) + QStringLiteral( "\n\n" ) + QCoreApplication::translate( "variable_help", "When used for calculations within the attribute table the row number will respect the original order of features from the underlying data source." ) + QStringLiteral( "\n\n" ) + QCoreApplication::translate( "variable_help", "When used from the field calculator the row numbering starts at 1, otherwise (e.g. from Processing tools) the row numbering starts from 0." ) );
```

nyalldawson committed:
Update src/core/expression/qgsexpression.cpp
